src/solvers/NelderMead.jl

```diff
@@ -1,6 +1,152 @@
+    NelderMeadState <: AbstractManoptSolverState
+
+Describes all parameters and the state of a Nealer-Mead heuristic based
+optimization algorithm.
+
+# Fields
+
+The naming of these parameters follows the [Wikipedia article](https://en.wikipedia.org/wiki/Nelder–Mead_method)
+of the Euclidean case. The default is given in brackets, the required value range
+after the description
+
+* `population` – an `Array{`point`,1}` of ``n+1`` points ``x_i``, ``i=1,…,n+1``, where ``n`` is the
+  dimension of the manifold.
+* `stopping_criterion` – ([`StopAfterIteration`](@ref)`(2000)`) a [`StoppingCriterion`](@ref)
+* `α` – (`1.`) reflection parameter (``α > 0``)
+* `γ` – (`2.`) expansion parameter (``γ > 0``)
+* `ρ` – (`1/2`) contraction parameter, ``0 < ρ ≤ \frac{1}{2}``,
+* `σ` – (`1/2`) shrink coefficient, ``0 < σ ≤ 1``
+* `p` – (`copy(population.pts[1])`) - a field to collect the current best value (initialized to _some_ point here)
+* `retraction_method` – (`default_retraction_method(M)`) the rectraction to use.
+* `inverse_retraction_method` - (`default_inverse_retraction_method(M)`) an inverse retraction to use.
+
+# Constructors
+
+    NelderMead(M[, population]; kwargs...)
+
+construct a Nelder-Mead Option with a default popultion (if not provided) of set of `dimension(M)+1` random points.
+
+In the constructor all fields (besides the population) are keyword arguments.
+"""
+mutable struct NelderMeadState{
+    T,
+    S<:StoppingCriterion,
+    Tα<:Real,
+    Tγ<:Real,
+    Tρ<:Real,
+    Tσ<:Real,
+    TR<:AbstractRetractionMethod,
+    TI<:AbstractInverseRetractionMethod,
+} <: AbstractManoptSolverState
+    population::NelderMeadSimplex{T}
+    stop::S
+    α::Tα
+    γ::Tγ
+    ρ::Tρ
+    σ::Tσ
+    p::T
+    costs::Vector{Float64}
+    retraction_method::TR
+    inverse_retraction_method::TI
+    function NelderMeadState(
+        M::AbstractManifold,
+        population::NelderMeadSimplex{T};
+        stopping_criterion::StoppingCriterion=StopAfterIteration(2000),
+        α=1.0,
+        γ=2.0,
+        ρ=1 / 2,
+        σ=1 / 2,
+        retraction_method::AbstractRetractionMethod=default_retraction_method(M),
+        inverse_retraction_method::AbstractInverseRetractionMethod=default_inverse_retraction_method(
+            M
+        ),
+        p::T=copy(M, population.pts[1]),
+    ) where {T}
+        return new{
+            T,
+            typeof(stopping_criterion),
+            typeof(α),
+            typeof(γ),
+            typeof(ρ),
+            typeof(σ),
+            typeof(retraction_method),
+            typeof(inverse_retraction_method),
+        }(
+            population,
+            stopping_criterion,
+            α,
+            γ,
+            ρ,
+            σ,
+            p,
+            [],
+            retraction_method,
+            inverse_retraction_method,
+        )
+    end
+end
+function NelderMeadState(M::AbstractManifold; kwargs...)
```

Comment:
does this method still work to do we have to turn this population into a NelderMeadSimplex first?

Reply:
ah yes as the tests (that currently fail) indicate we still would need the step via the constructor `NelderMeadState(M, population::Array{P})` or adapt tests and docs (and this function).

Reply:
Sorry, I didn't notice there is a second file with tests to check. I will adapt it so that population always needs to be passed by `NelderMeadSimplex`. What I don't like about `NelderMeadState(M, population::Array{P})` is that `population` could be confused with initial point.

Reply:
but population is not of type `P` as initial points are but an array of `P`, but since we are breaking change here anyways, choose what you prefer.

Reply:
What type `P`? Note that `NelderMeadSimplex` is parametrized by point type, not a vector of points type.

Reply:
For me there is no confusion of the `population::AbstractArray{P}` compared to a single initial point `p::P` since the array indicates it is an array of points and nor a single one

Reply:
(where `P` is the point type as usual in Manopt.jl)

Reply:
My confusion comes from the fact that `method_name(manifold, function, some_vector)` for many other method means "use `some_vector` as the starting point". `NelderMead(M, f, some_array)` could equally well mean `NelderMead(M, f, NelderMeadSimplex(some_array))` and `NelderMead(M, f, NelderMeadSimplex(M, some_array))` which have very different meanings.

Reply:
oh, well, the subtle part is that its not `some_vector` usually but `p::P` (not even unified to that) compared to NelderMead where it is `population::AbstractArray{P}` for me that would be fine – but if you feel that with the simplex type it is more clearer, that is also fine with me; though I do not see why the first one would not be clear.

Reply:
That would be fine if you could infer `P` from somewhere else but here `P` is inferred from either population or the initial point. We could validate that what we get is indeed a correct population but wrapping it in a simplex type seems easier to me.

Reply:
no problem, let's wrap it - jus be careful to update the documentation and also add doc-strings to the `NelderMeadSimplex` constructor(s)-